enhancements/sdk-external-and-pluggable-validations.md

```diff
@@ -0,0 +1,494 @@
+stdout.
+
+An example POC that takes an existing validator and outputs `ManifestErrors` can
+be found at [validator-poc][validator-poc1]
+
+Another example of a migration can be find at [ocp-olm-catalog-validator][camila-poc].
+This particular example does NOT yet output `ManifestErrors` format.
+
+TODO: do we need to add `json` tags to these structs?
+https://github.com/operator-framework/api/blob/master/pkg/validation/errors/error.go#L9-L16
+
+
+#### CLI
+
+The big question at hand is how do we indicate to the `operator-sdk` CLI that
+we want to run an external validator? Today, the `bundle validate` command takes
+in a few flags, here we will discuss how each might need to be changed to work
+with external validators.
+
+```
+Usage:
+  operator-sdk bundle validate [flags]
+
+...
+
+Flags:
+  -h, --help                                                 help for validate
+  -b, --image-builder string                                 Tool to pull and unpack bundle images. Only used when validating a bundle image. One of: [docker, podman, none] (default "docker")
+      --list-optional                                        List all optional validators available. When set, no validators will be run
+      --optional-values --optional-values=k8s-version=1.22   Inform a []string map of key=values which can be used by the validator. e.g. to check the operator bundle against an Kubernetes version that it is intended to be distributed use --optional-values=k8s-version=1.22 (default [])
+  -o, --output string                                        Result format for results. One of: [text, json-alpha1]. Note: output format types containing "alphaX" are subject to change and not covered by guarantees of stable APIs. (default "text")
+      --select-optional string                               Label selector to select optional validators to run. Run this command with '--list-optional' to list available optional validators
+
+Global Flags:
+      --plugins strings   plugin keys to be used for this subcommand execution
+      --verbose           Enable verbose logging
+```
+
+* *--help* works as is
+* *--image-builder* works as is
+* *--list-optional* would need to be updated to locate external validators.
+
+  TODO: How do we locate these validators? Do we use the XDG_CONFIG like Phase
+  2? Or allow users to pass in the location of the validators?
+
+* *--optional-values* would continue to be passed to the validators
+* *--output* indicates how we want to output the results.
+
+  TODO: what types do we want to support?
+
+* *--select-optional* takes in a label selector to specify which optional test
+  to run. We could allow a new label to indicate the location of the executable
+  to run as the validator. i.e. `validator-path=/path/to/validator/the-executable`
+
+TODO: need to define what the CLI for operator-sdk will look like. Are there
+any new flags? Do we use the environment variable?
```

Comment:
We could use viper which allows flags and env vars: https://github.com/spf13/viper
That would be aligned with Kubebuilder plan: https://github.com/kubernetes-sigs/kubebuilder/issues/2388#issuecomment-945004547

Reply:
I decided to avoid the environment variable altogether. I went with an alpha level flag: `--alpha-select-external`